File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Statement/CloseFileStatement.java
package Model.Statement;

import Model.Expression.ExpressionException;
import Model.Expression.IExpression;
import Model.State.ProgramState;
import Util.Dictionary.MyIDictionary;
import Util.Tuple.Tuple;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseFileStatement implements IStatement {
    private IExpression expFileId;

    public CloseFileStatement(IExpression expFileId) {
        this.expFileId = expFileId;
    }

    @Override
    public ProgramState execute(ProgramState programState) throws ExpressionException, StatementException {
        MyIDictionary<String,Integer> symbolTable = programState.getSymbolTable();
        int fileId = expFileId.evaluate(symbolTable);

        Tuple<String, BufferedReader> tuple = programState.getFileTable().get(fileId);

        if( tuple == null) {
            throw new StatementException("No entry in the file table for the given value!");
        }

        // close BufferedReader
        BufferedReader bufferedReader = tuple.getSecond();

        try {
            bufferedReader.close();
        }catch(IOException ex){
            throw new StatementException("Could not close file!");
        }

        // delete entry
        programState.getFileTable().remove(fileId);

        return programState;
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Statement/CloseFileStatement.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguage.Model.Expression;
using ToyLanguage.Model.State;
using ToyLanguage.Util.Dictionary;

namespace ToyLanguage.Model.Statement
{
    class CloseFileStatement : IStatement
    {
        private readonly IExpression expFileId;

        public CloseFileStatement(IExpression expFileId)
        {
            this.expFileId = expFileId;
        }
        public ProgramState Execute(ProgramState programState)
        {
            MyIDictionary<string, int> symbolTable = programState.SymbolTable;

            int fileDescriptor = expFileId.Evaluate(symbolTable);

            if (!programState.FileTable.TryGetValue(fileDescriptor, out Tuple<string, TextReader> value))
            {
                throw new StatementException("Invalid file id provided!");
            }

            TextReader reader = value.Item2;

            try
            {
                reader.Close();
            }
            catch (Exception)
            {
                throw new StatementException("Error while closing file!");
            }

            programState.FileTable.Remove(fileDescriptor);

            return programState;
        }
    }
}
